### ISSUE TITLE
Player and general opinion

### ISSUE THREAD

**Issue by jwiesler** [jpochyla/psst#6]
Hey, I really like your project. I'd like to collaborate but the roudmap you gave is way too general for me to be able to just pick and do something. Anyways, I want to share my experience using the player (I'm on Windows btw, no idea if those are specific bugs):
- It's really slow to use. Are you waiting for the web confirmation to change the play state? The music keeps running for about 2s after I click pause and seeking takes ages.
- The pause button stays a pause button even after pausing the track until I click it again, which changes it to the play button but does not start playing the song until I click a second time on it.
- Starting songs rapidly in succession will start playing each one of them for a brief moment after they finally loaded and not only the song that is currently selected at that point.
 
General personal opinion (debatable):
- Single clicking songs in playlists should not start playing them (at least not if clicking into the previously unfocused window).
- The player should be bigger and not 50:50 with the title (nothing has a title this long).
- All in all the player feels really out of sync with the gui. Maybe show changes only once they are actually in effect?
- The currently selected playlist should be highlighted like "Home" and "Library".
- The search bar looks a bit out of place.
- Otherwise I love the gui, it looks great and is intuitive.

**Comment by jpochyla** [comment]
Thanks for being interested! I agree that the roadmap is quite general, we can create issues for points that need further picking apart.

- The playback state reflected in the UI is indeed asynchronous to the playback thread, but it's just a cross-thread communication. Sound should pause immediately after clicking the pause button, and it does, in my tests. If you feel up to it, please debug.
- If the `Stopped -> Loading -> Playing` progression is not reflected immediately, I suspect it's a problem with Druid command delivery, which I sometimes encounter as well, but didn't yet have a chance to take a look at.
- That said, I've recently added the loudness spectrum visualisation for the seekbar, and because it's not down-sampled in any way, for longer tracks it takes too long to paint.

Re. the UI:

> Single clicking songs in playlists should not start playing them (at least not if clicking into the previously unfocused window).

I rather like the current behaviour! I know it's different than the official client, but as we don't have a concept of a selected track yet, I think that the most intuitive action for a single click is to play the track. That said, click-to-unfocused window is a platform-specific thing, I'm not familiar with how it usually works on Windows, and we can change the behaviour in such case.

> The player should be bigger and not 50:50 with the title (nothing has a title this long).

Some tracks do, although I agree that we do not need to display the title in full. Would you prefer a longer, more precise seekbar?

> All in all the player feels really out of sync with the gui. Maybe show changes only once they are actually in effect?

That's definitely not the intended behaviour.

> The currently selected playlist should be highlighted like "Home" and "Library".

👍 

> The search bar looks a bit out of place.

The search input is a really MVP at this point. I'd like to:
- Add a magnifying glass icon.
- Navigate to the results immediately after focusing the input.
- Search automatically after delay.
- Unfocus (and possibly clear) and go back after pressing escape.
- Unfocus and stay at the results after pressing enter.

> Otherwise I love the gui, it looks great and is intuitive.

Glad to hear! All in all, the UI is mostly the easiest thing that proves the concept, and would really benefit some further rethinking.

**Comment by JojiiOfficial** [comment]
For me the player works fine. Yes there is a very small delay between the UI and spotify but since the songs need to be loaded first and the request need to be made, I don't think this is really a bug from the player itself (I'm on arch btw).

I like this project too!

One thing I've discovered though is that non latin characters (Like japanese or korean song/album names) are displayed as empty boxes (prob. UTF-8 issues)...

**Comment by jpochyla** [comment]
> Yes there is a very small delay between the UI and the audio playback but since the songs need to be loaded first and the request need to be made, I don't think this is really a bug from the player itself

It is still a tiny bit slower than optimum (`minivorbis` is doing a read from the end of the file for some metadata that is not _strictly_ needed), but yes -- there might a be network delay.

> One thing I've discovered though is that non latin characters (Like japanese or korean song/album names) are displayed as empty boxes (prob. UTF-8 issues)...

Unfortunately I think this is the current state of the text stack for Druid on Linux.

**Comment by jwiesler** [comment]
Ok I compiled it in a Ubuntu VM and didn't find the big delay problems and button icon issue there, so it seems to be a Windows issue.

But I found another issue there: Seeking a paused song does not update the progress bar until I resume playing it again.

**Comment by jpochyla** [comment]
> Ok I compiled it in a Ubuntu VM and didn't find the big delay problems and button icon issue there, so it seems to be a Windows issue.

I suspect this to be related to `libsoundio`, which just got removed from master, so I hope it should help. It also should be easier to compile now. 

> But I found another issue there: Seeking a paused song does not update the progress bar until I resume playing it again.

Yes the progress report is updated only when pulling sound samples. One of the papercuts for now.

> And another thing is the menu bar. On MacOs it seems to be integrated into the task bar or whatever it is called there, but on Linux (where it could be too) and on Windows it is directly under the title bar. I don't know why, but i dont really like seeing menu bars that contain thinks like copy/paste that I won't ever use. Just my personal opinion, I don't know what can be done about this on Windows except just not showing it.

Yes I'd prefer not showing the menubar at all on Windows.

I think that both of these would deserve their own issues.

**Comment by jpochyla** [comment]
> But I found another issue there: Seeking a paused song does not update the progress bar until I resume playing it again.

Fixed in d136d3b

**Comment by jwiesler** [comment]
I don't know what you changed, but the problem with the pause button and the huge delay are gone since the last commit.

I'm able to deadlock the player in some form by spamming seek and play/pause. I'll update if I find out something more precise.

**Comment by jpochyla** [comment]
> I don't know what you changed, but the problem with the pause button and the huge delay are gone since the last commit. 

Good to hear! It's related to the switch to `miniaudio` backend.

> I'm able to deadlock the player in some form by spamming seek and play/pause.

Please open a separate issue.

> Starting new songs sporadically fails with OV_ENOTVORBIS, maybe you can do something with this information, I can't.

Also a separate issue, ideally with the track ID (right click, Copy Link).

**Comment by jpochyla** [comment]
This would be better suited for a discussion, thanks!
